File: python/infinicore/ops/moe_topk_vendor.py
from infinicore.dtype import float32, int32
from infinicore.lib import _infinicore
from infinicore.tensor import empty


def _alloc_outputs(gating_output, topk, topk_ids_dtype):
    shape = (gating_output.shape[0], topk)
    # The Iluvatar vendor extension allows fp32 weights for all gating dtypes; use fp32 by default for stable comparisons.
    topk_weights = empty(shape, dtype=float32, device=gating_output.device)
    topk_ids = empty(shape, dtype=topk_ids_dtype, device=gating_output.device)
    token_expert_indices = empty(shape, dtype=int32, device=gating_output.device)
    return topk_weights, topk_ids, token_expert_indices


def moe_topk_softmax_vendor(
    gating_output,
    topk,
    renormalize=False,
    correction_bias=None,
    *,
    out=None,
    topk_ids_dtype=int32,
):
    if out is None:
        out = _alloc_outputs(gating_output, topk, topk_ids_dtype)
    topk_weights, topk_ids, token_expert_indices = out
    _infinicore.moe_topk_softmax_vendor_(
        topk_weights._underlying,
        topk_ids._underlying,
        token_expert_indices._underlying,
        gating_output._underlying,
        renormalize,
        None if correction_bias is None else correction_bias._underlying,
    )
    return out


def moe_topk_sigmoid_vendor(
    gating_output,
    topk,
    renormalize=False,
    correction_bias=None,
    *,
    out=None,
    topk_ids_dtype=int32,
):
    if out is None:
        out = _alloc_outputs(gating_output, topk, topk_ids_dtype)
    topk_weights, topk_ids, token_expert_indices = out
    _infinicore.moe_topk_sigmoid_vendor_(
        topk_weights._underlying,
        topk_ids._underlying,
        token_expert_indices._underlying,
        gating_output._underlying,
        renormalize,
        None if correction_bias is None else correction_bias._underlying,
    )
    return out

File: src/infinicore/ops/moe_topk_vendor/moe_topk_vendor.cc
#include "infinicore/ops/moe_topk_vendor.hpp"
#include "../../utils.hpp"
#include "../vendor_ops/vendor_ops_dispatch.hpp"
#include <stdexcept>
namespace infinicore::op {
namespace {
void validate_vendor_topk(Tensor topk_weights, Tensor topk_ids, Tensor token_expert_indices, const Tensor &gating_output, const Tensor &correction_bias, const char *name) {
    if (!topk_weights || !topk_ids || !token_expert_indices || !gating_output) {
        throw std::runtime_error(std::string(name) + " expects non-empty tensors");
    }
    if (correction_bias) {
        INFINICORE_ASSERT_TENSORS_SAME_DEVICE(topk_weights, topk_ids, token_expert_indices, gating_output, correction_bias);
    } else {
        INFINICORE_ASSERT_TENSORS_SAME_DEVICE(topk_weights, topk_ids, token_expert_indices, gating_output);
    }
    if (gating_output->ndim() != 2 || topk_weights->ndim() != 2 || topk_ids->ndim() != 2 || token_expert_indices->ndim() != 2) {
        throw std::runtime_error(std::string(name) + " expects 2D tensors");
    }
    const auto tokens = gating_output->size(0), experts = gating_output->size(1), topk = topk_weights->size(1);
    if (topk_weights->size(0) != tokens || topk_ids->size(0) != tokens || token_expert_indices->size(0) != tokens || topk_ids->size(1) != topk || token_expert_indices->size(1) != topk) {
        throw std::runtime_error(std::string(name) + " expects output shapes (tokens, topk)");
    }
    if (topk < 1 || topk > 32 || topk > experts || experts < 1 || experts > 512) {
        throw std::runtime_error(std::string(name) + " supports experts in [1,512], topk in [1,32], topk<=experts");
    }
    if ((gating_output->dtype() == DataType::F16 || gating_output->dtype() == DataType::BF16) && (experts % 2) != 0) {
        throw std::runtime_error(std::string(name) + " requires even experts for fp16/bfloat16 gating");
    }
    if (gating_output->dtype() != DataType::F16 && gating_output->dtype() != DataType::BF16 && gating_output->dtype() != DataType::F32) {
        throw std::runtime_error(std::string(name) + " expects fp16/bfloat16/fp32 gating");
    }
    if (topk_weights->dtype() != DataType::F32 && topk_weights->dtype() != gating_output->dtype()) {
        throw std::runtime_error(std::string(name) + " expects topk_weights dtype float32 or same as gating");
    }
    if (topk_ids->dtype() != DataType::I32 && topk_ids->dtype() != DataType::I64) {
        throw std::runtime_error(std::string(name) + " expects topk_ids int32/int64");
    }
    if (token_expert_indices->dtype() != DataType::I32) {
        throw std::runtime_error(std::string(name) + " expects token_expert_indices int32");
    }
    if (correction_bias && (correction_bias->numel() != experts || correction_bias->dtype() != gating_output->dtype())) {
        throw std::runtime_error(std::string(name) + " expects correction_bias shape (experts,) and same dtype as gating");
    }
    if (!topk_weights->is_contiguous() || !topk_ids->is_contiguous() || !token_expert_indices->is_contiguous() || !gating_output->is_contiguous() || (correction_bias && !correction_bias->is_contiguous())) {
        throw std::runtime_error(std::string(name) + " expects contiguous tensors");
    }
}
} // namespace
void moe_topk_softmax_vendor_(Tensor topk_weights, Tensor topk_ids, Tensor token_expert_indices, const Tensor &gating_output, bool renormalize, const Tensor &correction_bias) {
    validate_vendor_topk(topk_weights, topk_ids, token_expert_indices, gating_output, correction_bias, "moe_topk_softmax_vendor");
    auto kernel = vendor_ops::lookup(
        vendor_ops::moe_topk_softmax_dispatcher(),
        gating_output->device().getType(),
        "moe_topk_softmax_vendor");
    kernel(topk_weights, topk_ids, token_expert_indices, gating_output, renormalize, correction_bias);
}
void moe_topk_sigmoid_vendor_(Tensor topk_weights, Tensor topk_ids, Tensor token_expert_indices, const Tensor &gating_output, bool renormalize, const Tensor &correction_bias) {
    validate_vendor_topk(topk_weights, topk_ids, token_expert_indices, gating_output, correction_bias, "moe_topk_sigmoid_vendor");
    auto kernel = vendor_ops::lookup(
        vendor_ops::moe_topk_sigmoid_dispatcher(),
        gating_output->device().getType(),
        "moe_topk_sigmoid_vendor");
    kernel(topk_weights, topk_ids, token_expert_indices, gating_output, renormalize, correction_bias);
}
} // namespace infinicore::op
